File: frontend/js/views/allowance_doubtful.js
// Allowance for doubtful accounts generator — aging-method bad-debt reserve and
// net realizable AR, via /calc/allowance-doubtful.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
const pct = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 }) + '%');
let LAST_DOC = null;

const SEED = [
    { label: 'Current', balance: 100000, rate: 1 },
    { label: '31-60', balance: 50000, rate: 5 },
    { label: '61-90', balance: 30000, rate: 20 },
    { label: '90+', balance: 20000, rate: 50 },
];

function rowHtml(tr) {
    return `
        <div class="mpb-row nec-row">
            <input type="text" class="adt-label" placeholder="${esc(t('view.adt.ph.label'))}" value="${esc(tr.label || '')}">
            <input type="number" step="1000" min="0" class="adt-balance" value="${tr.balance}">
            <input type="number" step="0.1" min="0" class="adt-rate" value="${tr.rate}">
            <button type="button" class="adt-del" data-i18n="view.adt.remove">Remove</button>
        </div>`;
}

export async function renderAllowanceDoubtful(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.adt.h1.title">// ALLOWANCE FOR DOUBTFUL ACCOUNTS</span></h1>
        <p class="muted small" data-i18n="view.adt.hint.intro">
            The aging-method estimate of uncollectible receivables. Each AR aging tier carries an estimated
            uncollectible percentage; the allowance is the sum of each tier's balance times its percentage, and
            net realizable AR is the total less the allowance. It also computes the adjusting entry to raise or
            lower an existing allowance balance. Drafting aid, not accounting advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.adt.h2.inputs">Reserve details</h2>
            <form id="adt-form" class="inline-form">
                <label><span data-i18n="view.adt.label.company">Company</span>
                    <input type="text" name="company_name" value="Acme Supply" required></label>
                <label><span data-i18n="view.adt.label.date">As-of date</span>
                    <input type="date" name="as_of_date" value="2026-06-30" required></label>
                <label><span data-i18n="view.adt.label.existing">Existing allowance ($)</span>
                    <input type="number" step="1000" min="0" name="existing_allowance_usd" value="12000"></label>
                <label><span data-i18n="view.adt.label.note">Note (optional)</span>
                    <input type="text" name="note" value="" placeholder="${esc(t('view.adt.ph.note'))}"></label>
            </form>
            <div class="mpb-head nec-head">
                <span data-i18n="view.adt.col.label">Aging tier</span>
                <span data-i18n="view.adt.col.balance">Balance ($)</span>
                <span data-i18n="view.adt.col.rate">Uncollectible %</span>
                <span></span>
            </div>
            <div id="adt-rows">${SEED.map(rowHtml).join('')}</div>
            <button type="button" id="adt-add" class="secondary" data-i18n="view.adt.add">+ Add tier</button>
        </div>
        <div id="adt-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#adt-form');
    const rowsEl = mount.querySelector('#adt-rows');

    const generate = async () => {
        const fd = new FormData(form);
        const tiers = [...rowsEl.querySelectorAll('.nec-row')].map((r) => ({
            label: (r.querySelector('.adt-label').value || '').trim(),
            balance_usd: Number(r.querySelector('.adt-balance').value) || 0,
            uncollectible_pct: Number(r.querySelector('.adt-rate').value) || 0,
        })).filter((tr) => tr.label);
        const body = {
            company_name: (fd.get('company_name') || '').trim(),
            as_of_date: fd.get('as_of_date'),
            tiers,
            existing_allowance_usd: Number(fd.get('existing_allowance_usd')) || 0,
            note: (fd.get('note') || '').trim(),
        };
        try {
            const doc = await api.calcAllowanceDoubtful(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.adt.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);

    mount.querySelector('#adt-add').addEventListener('click', () => {
        rowsEl.insertAdjacentHTML('beforeend', rowHtml({ label: '', balance: 0, rate: 0 }));
        applyUiI18n(rowsEl.lastElementChild);
        generate();
    });
    rowsEl.addEventListener('click', (e) => {
        if (e.target.classList.contains('adt-del')) {
            e.target.closest('.nec-row').remove();
            generate();
        }
    });
    form.addEventListener('input', () => { live(); });
    rowsEl.addEventListener('input', () => { live(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase(), ''];
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const rows = doc.rows.map((r) => `
        <tr><td>${esc(r.label)}</td><td>${money(r.balance_usd)}</td><td>${r.uncollectible_pct}%</td><td>${money(r.reserve_usd)}</td></tr>
    `).join('');
    const adj = doc.adjusting_entry_usd;
    const adjKey = adj >= 0 ? 'view.adt.card.entry_up' : 'view.adt.card.entry_down';
    const el = mount.querySelector('#adt-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card neg"><div class="label" data-i18n="view.adt.card.allowance">Allowance</div>
                    <div class="value">${money(doc.allowance_usd)}</div></div>
                <div class="card pos"><div class="label" data-i18n="view.adt.card.net">Net realizable AR</div>
                    <div class="value">${money(doc.net_realizable_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.adt.card.pct">% of AR</div>
                    <div class="value">${pct(doc.allowance_pct_of_ar)}</div></div>
                <div class="card"><div class="label" data-i18n="${adjKey}">Adjusting entry</div>
                    <div class="value">${money(Math.abs(adj))}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="adt-copy" type="button" data-i18n="view.adt.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="adt-download" type="button" data-i18n="view.adt.btn.download">Download .txt</button>
            </div>
        </div>
        <table class="data-table">
            <thead><tr>
                <th data-i18n="view.adt.th.label">Aging tier</th>
                <th data-i18n="view.adt.th.balance">Balance</th>
                <th data-i18n="view.adt.th.rate">Uncollectible</th>
                <th data-i18n="view.adt.th.reserve">Reserve</th>
            </tr></thead>
            <tbody>${rows}</tbody>
        </table>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#adt-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.adt.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.adt.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#adt-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'allowance-doubtful.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/allowance_doubtful.rs
//! Allowance for doubtful accounts (bad-debt reserve) — the aging-method estimate
//! of uncollectible receivables. Each accounts-receivable aging tier carries an
//! estimated uncollectible percentage; the allowance is the sum of each tier's
//! balance times its percentage, and net realizable AR is the total less the
//! allowance. Distinct from the statement of account, which only ages invoices;
//! this reserves against them. Drafting aid, not accounting advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct AgingTier {
    pub label: String,
    pub balance_usd: f64,
    /// Estimated uncollectible percentage for this tier.
    pub uncollectible_pct: f64,
}

#[derive(Debug, Clone, Deserialize)]
pub struct AllowanceInput {
    pub company_name: String,
    pub as_of_date: String,
    pub tiers: Vec<AgingTier>,
    /// Existing balance already in the allowance account (for the adjusting entry).
    #[serde(default)]
    pub existing_allowance_usd: f64,
    #[serde(default)]
    pub note: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct TierRow {
    pub label: String,
    pub balance_usd: f64,
    pub uncollectible_pct: f64,
    /// balance × pct.
    pub reserve_usd: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct AllowanceDoubtful {
    pub title: String,
    pub total_receivables_usd: f64,
    /// Sum of each tier's reserve — the required allowance balance.
    pub allowance_usd: f64,
    /// Total receivables − allowance.
    pub net_realizable_usd: f64,
    /// Allowance ÷ total receivables, percent.
    pub allowance_pct_of_ar: f64,
    /// Required allowance − existing balance (the adjusting entry to bad-debt expense).
    pub adjusting_entry_usd: f64,
    pub rows: Vec<TierRow>,
    pub clauses: Vec<DocClause>,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

pub fn generate(i: &AllowanceInput) -> AllowanceDoubtful {
    let rows: Vec<TierRow> = i
        .tiers
        .iter()
        .map(|t| TierRow {
            label: t.label.clone(),
            balance_usd: cents(t.balance_usd),
            uncollectible_pct: t.uncollectible_pct,
            reserve_usd: cents(t.balance_usd * t.uncollectible_pct / 100.0),
        })
        .collect();

    let total = cents(i.tiers.iter().map(|t| t.balance_usd).sum());
    let allowance = cents(rows.iter().map(|r| r.reserve_usd).sum());
    let net = cents(total - allowance);
    let allowance_pct = if total > 0.0 {
        cents(allowance / total * 100.0)
    } else {
        0.0
    };
    let adjusting = cents(allowance - i.existing_allowance_usd);

    let detail = if rows.is_empty() {
        "No receivables aging tiers listed.".to_string()
    } else {
        rows.iter()
            .map(|r| {
                format!(
                    "  • {}: {} × {:.2}% = {}",
                    r.label, money(r.balance_usd), r.uncollectible_pct, money(r.reserve_usd)
                )
            })
            .collect::<Vec<_>>()
            .join("\n")
    };

    let entry_desc = if adjusting >= 0.0 {
        format!(
            "Debit Bad Debt Expense and credit Allowance for Doubtful Accounts {} to raise the existing balance of {} to the required {}.",
            money(adjusting.abs()), money(i.existing_allowance_usd), money(allowance)
        )
    } else {
        format!(
            "Debit Allowance for Doubtful Accounts and credit Bad Debt Expense {} to reduce the existing balance of {} to the required {}.",
            money(adjusting.abs()), money(i.existing_allowance_usd), money(allowance)
        )
    };

    let mut clauses = vec![
        DocClause {
            heading: "Statement".into(),
            body: format!("Company: {}\nAs of: {}", i.company_name, i.as_of_date),
        },
        DocClause { heading: "Aging Reserve".into(), body: detail },
        DocClause {
            heading: "Allowance".into(),
            body: format!(
                "Total receivables of {} require an allowance of {} ({:.2}% of AR), leaving net realizable receivables of {}.",
                money(total), money(allowance), allowance_pct, money(net)
            ),
        },
        DocClause { heading: "Adjusting Entry".into(), body: entry_desc },
    ];

    let note = i.note.trim();
    if !note.is_empty() {
        clauses.push(DocClause { heading: "Note".into(), body: note.to_string() });
    }

    AllowanceDoubtful {
        title: "Allowance for Doubtful Accounts".into(),
        total_receivables_usd: total,
        allowance_usd: allowance,
        net_realizable_usd: net,
        allowance_pct_of_ar: allowance_pct,
        adjusting_entry_usd: adjusting,
        rows,
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn tier(label: &str, bal: f64, pct: f64) -> AgingTier {
        AgingTier { label: label.into(), balance_usd: bal, uncollectible_pct: pct }
    }

    fn base() -> AllowanceInput {
        AllowanceInput {
            company_name: "Acme Supply".into(),
            as_of_date: "2026-06-30".into(),
            tiers: vec![
                tier("Current", 100_000.0, 1.0),
                tier("31-60", 50_000.0, 5.0),
                tier("61-90", 30_000.0, 20.0),
                tier("90+", 20_000.0, 50.0),
            ],
            existing_allowance_usd: 12_000.0,
            note: String::new(),
        }
    }

    #[test]
    fn allowance_and_net_realizable() {
        let d = generate(&base());
        assert!(close(d.total_receivables_usd, 200_000.0));
        // 1,000 + 2,500 + 6,000 + 10,000 = 19,500.
        assert!(close(d.allowance_usd, 19_500.0));
        assert!(close(d.net_realizable_usd, 180_500.0));
        assert!(close(d.allowance_pct_of_ar, 9.75));
    }

    #[test]
    fn adjusting_entry_raises_to_required() {
        let d = generate(&base());
        // 19,500 required − 12,000 existing = 7,500.
        assert!(close(d.adjusting_entry_usd, 7_500.0));
    }

    #[test]
    fn over_reserved_negative_entry() {
        let d = generate(&AllowanceInput { existing_allowance_usd: 25_000.0, ..base() });
        // 19,500 − 25,000 = −5,500 (reduce the allowance).
        assert!(close(d.adjusting_entry_usd, -5_500.0));
        assert!(d.clauses.iter().any(|c| c.body.contains("reduce")));
    }

    #[test]
    fn per_tier_reserves() {
        let d = generate(&base());
        assert_eq!(d.rows.len(), 4);
        assert!(close(d.rows[3].reserve_usd, 10_000.0));
    }

    #[test]
    fn empty_tiers_zero() {
        let d = generate(&AllowanceInput { tiers: vec![], ..base() });
        assert!(close(d.allowance_usd, 0.0));
        assert!(d.clauses.iter().any(|c| c.body.contains("No receivables")));
    }

    #[test]
    fn note_appended_when_present() {
        let d = generate(&AllowanceInput { note: "Per credit committee.".into(), ..base() });
        assert!(d.clauses.iter().any(|c| c.heading == "Note" && c.body.contains("credit committee")));
    }
}
